starknet_py/proxy/contract_abi_resolver.py

```diff
@@ -0,0 +1,157 @@
+import warnings
+from enum import Enum
+from typing import List, TypedDict, Tuple
+
+from starknet_py.net.client import Client
+from starknet_py.net.client_errors import ContractNotFoundError, ClientError
+from starknet_py.net.client_models import Abi, DeclaredContract
+from starknet_py.net.models import Address
+from starknet_py.proxy.proxy_check import (
+    OpenZeppelinProxyCheck,
+    ArgentProxyCheck,
+    ProxyCheck,
+)
+
+
+class ProxyConfig(TypedDict, total=False):
+    """
+    Proxy resolving configuration
+    """
+
+    proxy_checks: List[ProxyCheck]
+    """
+    List of classes implementing :class:`starknet_py.proxy.proxy_check.ProxyCheck` ABC,
+    that will be used for checking if contract at the address is a proxy contract.
+    """
+
+
+def prepare_proxy_config(proxy_config: ProxyConfig) -> ProxyConfig:
+    if "max_steps" in proxy_config:
+        warnings.warn(
+            "ProxyConfig.max_steps is deprecated. Contract.from_address always makes at most 1 step.",
+            category=DeprecationWarning,
+        )
+    if "proxy_checks" in proxy_config:
+        return proxy_config
+
+    proxy_checks = [OpenZeppelinProxyCheck(), ArgentProxyCheck()]
+    return {"proxy_checks": proxy_checks}
+
+
+class ImplementationType(Enum):
+    """
+    Enum representing transaction types
+    """
+
+    CLASS_HASH = "class_hash"
+    ADDRESS = "address"
+
+
+class ContractAbiResolver:
+    """
+    Class for resolving abi of a contract
+    """
+
+    def __init__(
+        self,
+        address: Address,
+        client: Client,
+        proxy_config: ProxyConfig,
```

Comment:
So if you don't want to resolve proxies at all, you have to pass `proxy_config=[]`? Maybe that's been already discussed but perhaps more ergonomic making it `Optional[ProxyConfig]=None` and defaulting to empty list if it's None

cc @Solpatium

Reply:
The most important aspect of ContractAbiResolver is being able to resolve ABI behind a proxy. I don't think making it easier is needed here. If users complain we can change it of course.

Reply:
Okay, let's keep it this way then.